MsCorePkg/Library/MuSecureBootKeySelectorLib/MuSecureBootKeySelectorLib.c

```diff
@@ -17,9 +17,228 @@
+    0xF6, 0x12, 0x4E, 0x34, 0x12, 0x5B, 0xEE, 0x3F, 0xE6, 0xD7, 0x9A, 0x57, 0x4E, 0xAA, 0x7B, 0x91,
+    0xC0, 0xE7, 0xBD, 0x9D, 0x92, 0x9C, 0x1A, 0x32, 0x11, 0x78, 0xEF, 0xD6, 0x11, 0xDA, 0xD9, 0x01
+  },
+  { // Microsoft Windows First Party Certificate 2023
+    0x07, 0x6F, 0x1F, 0xEA, 0x90, 0xAC, 0x29, 0x15, 0x5E, 0xBF, 0x77, 0xC1, 0x76, 0x82, 0xF7, 0x5F,
+    0x1F, 0xDD, 0x1B, 0xE1, 0x96, 0xDA, 0x30, 0x2D, 0xC8, 0x46, 0x1E, 0x35, 0x0A, 0x9A, 0xE3, 0x30
+  }
+};
+
+//
+// This is the handle for the Hash2ServiceBinding Protocol instance this driver produces
+// if the platform does not provide one.
+//
+EFI_HANDLE  mHash2ServiceHandle = NULL;
+
+/**
+  Determine if the Secure Boot Keys in the provided database (DbVar) correspond to
+  Microsoft-serviced keys.
+
+  This function checks the certificates in the provided Secure Boot database (DbVar)
+  against a predefined list of Microsoft certificates.
+  If all certificates in the database match the Microsoft-serviced certificates,
+  the machine is considered `Microsoft Serviced.` If any certificate does not match,
+  the machine is considered `CUSTOM,` and the function returns an error.
+
+  @param[in]  DbVar               Pointer to the Secure Boot database (db) variable.
+  @param[in]  DbVarSize           Size of the Secure Boot database in bytes.
+  @param[out] IsMsServicedConfig  Pointer to a BOOLEAN that will be set to TRUE if
+                                  the database matches Microsoft-serviced keys, or
+                                  FALSE otherwise.
+
+  @retval EFI_SUCCESS             The operation completed successfully, and the result
+                                  is stored in IsMsServicedConfig.
+  @retval EFI_NOT_FOUND           A certificate in the database does not match any
+                                  Microsoft-serviced certificates.
+  @retval EFI_UNSUPPORTED         The hashing protocol or service binding protocol
+                                  required for the operation is not available.
+  @retval EFI_OUT_OF_RESOURCES    Memory allocation failed during the operation.
+  @retval Other                   An error occurred during the operation.
+
+**/
+EFI_STATUS
+EFIAPI
+IsMsftServicedConfig (
+  UINT8    *DbVar,
+  UINTN    DbVarSize,
+  BOOLEAN  *IsMsServicedConfig
+  )
+{
+  EFI_STATUS                    Status;
+  EFI_SIGNATURE_LIST            *CertList;
+  EFI_SIGNATURE_DATA            *Cert;
+  UINTN                         CertCount;
+  UINT8                         *RootCert;
+  UINTN                         RootCertSize;
+  UINTN                         Index;
+  UINTN                         CertIndex;
+  EFI_SERVICE_BINDING_PROTOCOL  *Hash2ServiceBinding;
+  EFI_HASH2_PROTOCOL            *Hash2Protocol;
+  EFI_HASH2_OUTPUT              HashOutput;
+
+  if (IsMsServicedConfig != NULL) {
+    *IsMsServicedConfig = FALSE;
+  }
+
+  Status = gBS->LocateProtocol (&gEfiHash2ProtocolGuid, NULL, (VOID **)&Hash2Protocol);

```

Comment:
If the platform publishes `gEfiHash2ServiceBindingProtocolGuid` that means they're expecting you to use it create a hash protocol child instance for use here, so shouldn't that be checked first and then fall back on a default `gEfiHash2ProtocolGuid` if it's not available?